JDK8 Mauve[gnu.testlet.java.beans.PropertyDescriptor.constructorTest2] - IllegalArgumentException: Can not copy a non-root Method

Failure link
------------

From an internal build `job/Test_openjdk8_j9_special.system_x86-64_mac_testList_2/109/`(`osxrt1`):
```
openjdk version "1.8.0_322"
IBM Semeru Runtime Open Edition (build 1.8.0_322-b04)
Eclipse OpenJ9 VM (build master-2871b64e3, JRE 1.8.0 Mac OS X amd64-64-Bit Compressed References 20211210_266 (JIT enabled, AOT enabled)
OpenJ9   - 2871b64e3
OMR      - 5a958d708
JCL      - 854c9aec based on jdk8u322-b04)
```
[Rerun in Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?SDK_RESOURCE=upstream&amp;TARGET=MauveSingleInvocLoad_special_5m_13&amp;TEST_FLAG=&amp;UPSTREAM_TEST_JOB_NAME=Test_openjdk8_j9_special.system_x86-64_mac&amp;DOCKER_REQUIRED=false&amp;ACTIVE_NODE_TIMEOUT=0&amp;VENDOR_TEST_DIRS=&amp;EXTRA_DOCKER_ARGS=&amp;TKG_OWNER_BRANCH=AdoptOpenJDK%3Amaster&amp;TEST_PARALLELIZATION_PARAMS=&amp;OPENJ9_SYSTEMTEST_OWNER_BRANCH=eclipse%3Amaster&amp;PLATFORM=x86-64_mac&amp;GENERATE_JOBS=true&amp;KEEP_REPORTDIR=false&amp;PERSONAL_BUILD=false&amp;ADOPTOPENJDK_REPO=https%3A%2F%2Fgithub.com%2Fadoptium%2Faqa-tests.git&amp;LABEL=&amp;TEST_OPTIONS_PARAMS=&amp;EXTRA_OPTIONS=&amp;CUSTOMIZED_SDK_URL=+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk8u%2Fjdk8u-mac-x64-openj9%2F266%2Fibm-semeru-open-debugimage_x64_mac_JDK8U_2021-12-11-03-31.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk8u%2Fjdk8u-mac-x64-openj9%2F266%2Fibm-semeru-open-jre_x64_mac_JDK8U_2021-12-11-03-31.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk8u%2Fjdk8u-mac-x64-openj9%2F266%2Fibm-semeru-open-jdk_x64_mac_JDK8U_2021-12-11-03-31.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk8u%2Fjdk8u-mac-x64-openj9%2F266%2Fibm-semeru-open-testimage_x64_mac_JDK8U_2021-12-11-03-31.tar.gz&amp;BUILD_IDENTIFIER=&amp;NON_AQA_TEST_REPOS_HELP_TEXT=&amp;ADOPTOPENJDK_BRANCH=master&amp;LIGHT_WEIGHT_CHECKOUT=true&amp;NON_AQA_TEST_REPOS=&amp;ARTIFACTORY_SERVER=na.artifactory.swg-devops&amp;USE_JRE=false&amp;TEST_REPO_PARAMS=&amp;TEST_SELECTION_PARAMS=&amp;TEST_PARALLELIZATION_PARAMS_HELP_TEXT=&amp;JDK_SELECTION_PARAMS=&amp;KEEP_WORKSPACE=false&amp;USER_CREDENTIALS_ID=&amp;JDK_VERSION=8&amp;ITERATIONS=1&amp;VENDOR_TEST_REPOS=&amp;JDK_REPO=https%3A%2F%2Fgithub.com%2Fibmruntimes%2Fopenj9-openjdk-jdk8&amp;PLATFORM_AND_MACHINE_HELP_TEXT=&amp;RELEASE_TAG=&amp;OPENJ9_BRANCH=master&amp;OPENJ9_SHA=&amp;JCK_GIT_REPO=&amp;VENDOR_TEST_BRANCHES=&amp;OPENJ9_REPO=https%3A%2F%2Fgithub.com%2Feclipse-openj9%2Fopenj9.git&amp;UPSTREAM_JOB_NAME=&amp;CLOUD_PROVIDER=&amp;PLATFORM_AND_MACHINE=&amp;CUSTOM_TARGET=&amp;VENDOR_TEST_SHAS=&amp;JDK_BRANCH=openj9&amp;LABEL_ADDITION=&amp;ARTIFACTORY_REPO=&amp;ARTIFACTORY_ROOT_DIR=&amp;POST_RUN_PARAMS_HELP_TEXT=&amp;UPSTREAM_TEST_JOB_NUMBER=113&amp;DOCKERIMAGE_TAG=&amp;JDK_SELECTION_PARAMS_HELP_TEXT=&amp;JDK_IMPL=openj9&amp;TEST_TIME=&amp;SSH_AGENT_CREDENTIAL=83181e25-eea4-4f55-8b3e-e79615733226&amp;AUTO_DETECT=true&amp;SLACK_CHANNEL=&amp;DYNAMIC_COMPILE=false&amp;TKG_SHA=&amp;TEST_SELECTION_PARAMS_HELP_TEXT=&amp;ADOPTOPENJDK_SYSTEMTEST_OWNER_BRANCH=adoptium%3Amaster&amp;CUSTOMIZED_SDK_URL_CREDENTIAL_ID=4e18ffe7-b1b1-4272-9979-99769b68bcc2&amp;ARCHIVE_TEST_RESULTS=false&amp;NUM_MACHINES=&amp;OPENJDK_SHA=&amp;TRSS_URL=http%3A%2F%2Ftrss1.fyre.ibm.com&amp;BUILD_LIST=system&amp;UPSTREAM_JOB_NUMBER=&amp;STF_OWNER_BRANCH=adoptium%3Amaster&amp;TEST_REPO_PARAMS_HELP_TEXT=&amp;POST_RUN_PARAMS=&amp;TIME_LIMIT=20&amp;TEST_OPTIONS_PARAMS_HELP_TEXT=&amp;JVM_OPTIONS=&amp;PARALLEL=None) - Change TARGET to run only the failed test targets.

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------
```
[2021-12-11T16:04:11.395Z] variation: Mode553
[2021-12-11T16:04:11.395Z] JVM_OPTIONS:  -XX:+UseCompressedOops -Xgcpolicy:balanced -Xjit:count=0 

[2021-12-11T16:04:31.474Z] LT  11:04:30.233 - Starting thread. Suite=0 thread=0
[2021-12-11T16:04:53.325Z] LT  11:04:50.793 - Completed 6.9%. Number of tests started=2352
[2021-12-11T16:05:11.015Z] LT  16:05:09.680 - First failure detected by thread: load-0. Not creating dumps as no dump generation is requested for this load test
[2021-12-11T16:05:11.015Z] LT  16:05:09.689 - Test failed
[2021-12-11T16:05:11.015Z] LT    Failure num.  = 1
[2021-12-11T16:05:11.015Z] LT    Test number   = 547
[2021-12-11T16:05:11.015Z] LT    Test details  = 'Mauve[gnu.testlet.java.beans.PropertyDescriptor.constructorTest2]'
[2021-12-11T16:05:11.015Z] LT    Suite number  = 0
[2021-12-11T16:05:11.015Z] LT    Thread number = 0
[2021-12-11T16:05:11.015Z] LT  >>> Captured test output >>>

[2021-12-11T16:05:11.016Z] LT  Test failed:
[2021-12-11T16:05:11.016Z] LT  java.lang.InternalError: java.lang.reflect.InvocationTargetException
[2021-12-11T16:05:11.016Z] LT  	at java.lang.Class.newInternalError(Class.java:3980)
[2021-12-11T16:05:11.016Z] LT  	at java.lang.Class.copyMethods(Class.java:4162)
[2021-12-11T16:05:11.016Z] LT  	at java.lang.Class.cacheMethods(Class.java:4222)
[2021-12-11T16:05:11.016Z] LT  	at java.lang.Class.getMethods(Class.java:1630)
[2021-12-11T16:05:11.016Z] LT  	at java.beans.Introspector.getPublicDeclaredMethods(Introspector.java:1336)
[2021-12-11T16:05:11.016Z] LT  	at java.beans.Introspector.internalFindMethod(Introspector.java:1378)
[2021-12-11T16:05:11.016Z] LT  	at java.beans.Introspector.internalFindMethod(Introspector.java:1418)
[2021-12-11T16:05:11.016Z] LT  	at java.beans.Introspector.findMethod(Introspector.java:1450)
[2021-12-11T16:05:11.016Z] LT  	at java.beans.PropertyDescriptor.getWriteMethod(PropertyDescriptor.java:298)
[2021-12-11T16:05:11.016Z] LT  	at java.beans.PropertyDescriptor.<init>(PropertyDescriptor.java:110)
[2021-12-11T16:05:11.016Z] LT  	at java.beans.PropertyDescriptor.<init>(PropertyDescriptor.java:71)
[2021-12-11T16:05:11.016Z] LT  	at gnu.testlet.java.beans.PropertyDescriptor.constructorTest2.constructor1Tests(constructorTest2.java:600)
[2021-12-11T16:05:11.016Z] LT  	at gnu.testlet.java.beans.PropertyDescriptor.constructorTest2.test(constructorTest2.java:700)
[2021-12-11T16:05:11.016Z] LT  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2021-12-11T16:05:11.016Z] LT  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[2021-12-11T16:05:11.016Z] LT  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2021-12-11T16:05:11.016Z] LT  	at java.lang.reflect.Method.invoke(Method.java:498)
[2021-12-11T16:05:11.016Z] LT  	at net.adoptopenjdk.loadTest.adaptors.MauveAdaptor.executeTest(MauveAdaptor.java:74)
[2021-12-11T16:05:11.016Z] LT  	at net.adoptopenjdk.loadTest.LoadTestRunner$2.run(LoadTestRunner.java:182)
[2021-12-11T16:05:11.016Z] LT  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2021-12-11T16:05:11.016Z] LT  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2021-12-11T16:05:11.016Z] LT  	at java.lang.Thread.run(Thread.java:827)
[2021-12-11T16:05:11.016Z] LT  Caused by: java.lang.reflect.InvocationTargetException
[2021-12-11T16:05:11.016Z] LT  	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
[2021-12-11T16:05:11.016Z] LT  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2021-12-11T16:05:11.016Z] LT  	at java.lang.reflect.Method.invoke(Method.java:498)
[2021-12-11T16:05:11.016Z] LT  	at java.lang.Class.copyMethods(Class.java:4156)
[2021-12-11T16:05:11.016Z] LT  	... 20 more
[2021-12-11T16:05:11.016Z] LT  Caused by: java.lang.IllegalArgumentException: Can not copy a non-root Method
[2021-12-11T16:05:11.016Z] LT  	at java.lang.reflect.Method.copy(Method.java:151)
[2021-12-11T16:05:11.016Z] LT  	... 24 more
[2021-12-11T16:05:11.016Z] LT  <<<

[2021-12-11T16:09:30.458Z] MauveSingleInvocLoad_special_5m_13_FAILED
```

50x grinder at `job/Grinder/19982/`

[Internal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk8_j9_special.system_aarch64_linux_testList_4/343/)(`ub18-aarch64-4`)
```
openjdk version "1.8.0_362-beta"
IBM Semeru Runtime Open Edition (build 1.8.0_362-beta-202301170123-b07)
Eclipse OpenJ9 VM (build master-082682ae5, JRE 1.8.0 Linux aarch64-64-Bit Compressed References 20230117_540 (JIT enabled, AOT enabled)
OpenJ9   - 082682ae5
OMR      - ec81a72de
JCL      - cd84b623c based on jdk8u362-b07)

[2023-01-17T06:32:43.627Z] variation: Mode107-OSRG
[2023-01-17T06:32:43.627Z] JVM_OPTIONS:  -Xgcpolicy:optthruput -Xdebug -Xrunjdwp:transport=dt_socket,address=8888,server=y,onthrow=no.pkg.foo,launch=echo -Xjit:enableOSR,enableOSROnGuardFailure,count=1,disableAsyncCompilation 

[2023-01-17T06:37:35.976Z] LT  Test failed:
[2023-01-17T06:37:35.976Z] LT  java.lang.InternalError: java.lang.reflect.InvocationTargetException
[2023-01-17T06:37:35.976Z] LT  	at java.lang.Class.newInternalError(Class.java:3995)
[2023-01-17T06:37:35.976Z] LT  	at java.lang.Class.copyMethods(Class.java:4177)
[2023-01-17T06:37:35.976Z] LT  	at java.lang.Class.cacheMethods(Class.java:4237)
[2023-01-17T06:37:35.976Z] LT  	at java.lang.Class.getMethods(Class.java:1632)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.Introspector.getPublicDeclaredMethods(Introspector.java:1336)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.Introspector.internalFindMethod(Introspector.java:1378)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.Introspector.internalFindMethod(Introspector.java:1418)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.Introspector.findMethod(Introspector.java:1450)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.Introspector.findMethod(Introspector.java:1430)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.PropertyDescriptor.getReadMethod(PropertyDescriptor.java:229)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.PropertyDescriptor.<init>(PropertyDescriptor.java:106)
[2023-01-17T06:37:35.976Z] LT  	at java.beans.PropertyDescriptor.<init>(PropertyDescriptor.java:71)
[2023-01-17T06:37:35.976Z] LT  	at gnu.testlet.java.beans.PropertyDescriptor.constructorTest2.constructor1Tests(constructorTest2.java:508)
[2023-01-17T06:37:35.976Z] LT  	at gnu.testlet.java.beans.PropertyDescriptor.constructorTest2.test(constructorTest2.java:700)
[2023-01-17T06:37:35.976Z] LT  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2023-01-17T06:37:35.976Z] LT  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[2023-01-17T06:37:35.976Z] LT  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2023-01-17T06:37:35.976Z] LT  	at java.lang.reflect.Method.invoke(Method.java:498)
[2023-01-17T06:37:35.976Z] LT  	at net.adoptopenjdk.loadTest.adaptors.MauveAdaptor.executeTest(MauveAdaptor.java:74)
[2023-01-17T06:37:35.976Z] LT  	at net.adoptopenjdk.loadTest.LoadTestRunner$2.run(LoadTestRunner.java:182)
[2023-01-17T06:37:35.976Z] LT  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2023-01-17T06:37:35.976Z] LT  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[2023-01-17T06:37:35.976Z] LT  	at java.lang.Thread.run(Thread.java:826)
[2023-01-17T06:37:35.976Z] LT  Caused by: java.lang.reflect.InvocationTargetException
[2023-01-17T06:37:35.976Z] LT  	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
[2023-01-17T06:37:35.976Z] LT  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2023-01-17T06:37:35.976Z] LT  	at java.lang.reflect.Method.invoke(Method.java:498)
[2023-01-17T06:37:35.976Z] LT  	at java.lang.Class.copyMethods(Class.java:4171)
[2023-01-17T06:37:35.976Z] LT  	... 21 more
[2023-01-17T06:37:35.976Z] LT  Caused by: java.lang.IllegalArgumentException: Can not copy a non-root Method
[2023-01-17T06:37:35.976Z] LT  	at java.lang.reflect.Method.copy(Method.java:151)
[2023-01-17T06:37:35.976Z] LT  	... 25 more

[2023-01-17T06:38:11.025Z] MauveMultiThrdLoad_special_5m_23_FAILED
```

[50x grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/30469/)

https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_special.system_x86-64_windows_Release_testList_2/77
MauveSingleInvocLoad_special_5m_17
```
06:53:57  LT  10:53:54.657 - Test failed
06:53:57  LT    Failure num.  = 1
06:53:57  LT    Test number   = 546
06:53:57  LT    Test details  = 'Mauve[gnu.testlet.java.beans.PropertyDescriptor.constructorTest1]'
06:53:57  LT    Suite number  = 0
06:53:57  LT    Thread number = 0
06:53:57  LT  >>> Captured test output >>>
06:53:57  LT  Test failed:
06:53:57  LT  java.lang.InternalError: java.lang.reflect.InvocationTargetException
06:53:57  LT  	at java.lang.Class.newInternalError(Class.java:4010)
06:53:57  LT  	at java.lang.Class.copyMethods(Class.java:4192)
06:53:57  LT  	at java.lang.Class.cacheMethods(Class.java:4252)
06:53:57  LT  	at java.lang.Class.getMethods(Class.java:1652)
06:53:57  LT  	at java.beans.Introspector.getPublicDeclaredMethods(Introspector.java:1336)
06:53:57  LT  	at java.beans.Introspector.internalFindMethod(Introspector.java:1378)
06:53:57  LT  	at java.beans.Introspector.internalFindMethod(Introspector.java:1418)
06:53:57  LT  	at java.beans.Introspector.findMethod(Introspector.java:1450)
06:53:57  LT  	at java.beans.Introspector.findMethod(Introspector.java:1430)
06:53:57  LT  	at java.beans.PropertyDescriptor.getReadMethod(PropertyDescriptor.java:229)
06:53:57  LT  	at java.beans.PropertyDescriptor.<init>(PropertyDescriptor.java:106)
06:53:57  LT  	at java.beans.PropertyDescriptor.<init>(PropertyDescriptor.java:71)
06:53:57  LT  	at gnu.testlet.java.beans.PropertyDescriptor.constructorTest1.test(constructorTest1.java:49)
06:53:57  LT  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
06:53:57  LT  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
06:53:57  LT  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
06:53:57  LT  	at java.lang.reflect.Method.invoke(Method.java:503)
06:53:57  LT  	at net.adoptopenjdk.loadTest.adaptors.MauveAdaptor.executeTest(MauveAdaptor.java:74)
06:53:57  LT  	at net.adoptopenjdk.loadTest.LoadTestRunner$2.run(LoadTestRunner.java:182)
06:53:57  LT  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
06:53:57  LT  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
06:53:57  LT  	at java.lang.Thread.run(Thread.java:826)
06:53:57  LT  Caused by: java.lang.reflect.InvocationTargetException
06:53:57  LT  	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
06:53:57  LT  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
06:53:57  LT  	at java.lang.reflect.Method.invoke(Method.java:503)
06:53:57  LT  	at java.lang.Class.copyMethods(Class.java:4186)
06:53:57  LT  	... 20 more
06:53:57  LT  Caused by: java.lang.IllegalArgumentException: Can not copy a non-root Method
06:53:57  LT  	at java.lang.reflect.Method.copy(Method.java:156)
06:53:57  LT  	... 24 more
```

--- a/jcl/src/java.base/share/classes/java/lang/Class.java
+++ b/jcl/src/java.base/share/classes/java/lang/Class.java
@@ -1409,7 +1409,7 @@ private List<Method> cacheDeclaredPublicMethods(List<Method> methods, CacheKey c
 				methods.set(i, methodPut);
 			}
 		}
-		cache.insert(cacheKey, methods);
+		cache.insertRoot(cacheKey, methods);
 	} finally {
 		if (cache != null) {
 			cache.release();
@@ -4549,19 +4549,23 @@ private String getParameterTypesSignature(boolean throwException, String name, C
 	return signature.toString();
 }
 
+/*[IF JAVA_SPEC_VERSION <= 11] */
+private static Method getRootMethod;
 /*[IF JAVA_SPEC_VERSION == 8]*/
 /*[PR CMVC 114820, CMVC 115873, CMVC 116166] add reflection cache */
 private static Method copyMethod, copyField, copyConstructor;
+private static Field rootField;
 private static Field methodParameterTypesField;
 private static Field constructorParameterTypesField;
-private static final Object[] NoArgs = new Object[0];
 /*[ENDIF] JAVA_SPEC_VERSION == 8 */
+private static final Object[] NoArgs = new Object[0];
+/*[ENDIF] JAVA_SPEC_VERSION <= 11 */
 
 /*[PR JAZZ 107786] constructorParameterTypesField should be initialized regardless of reflectCacheEnabled or not */
 static void initCacheIds(boolean cacheEnabled, boolean cacheDebug) {
 	reflectCacheEnabled = cacheEnabled;
 	reflectCacheDebug = cacheDebug;
-	/*[IF JAVA_SPEC_VERSION == 8]*/
+	/*[IF JAVA_SPEC_VERSION <= 11]*/
 	AccessController.doPrivileged(new PrivilegedAction<Void>() {
 		@Override
 		public Void run() {
@@ -4569,12 +4573,14 @@ static void initCacheIds(boolean cacheEnabled, boolean cacheDebug) {
 			return null;
 		}
 	});
-	/*[ENDIF] JAVA_SPEC_VERSION == 8 */
+	/*[ENDIF] JAVA_SPEC_VERSION <= 11 */
 }
+
 static void setReflectCacheAppOnly(boolean cacheAppOnly) {
 	reflectCacheAppOnly = cacheAppOnly;
 }
-/*[IF JAVA_SPEC_VERSION == 8]*/
+
+/*[IF JAVA_SPEC_VERSION <= 11]*/
 @SuppressWarnings("nls")
 static void doInitCacheIds() {
 	/*
@@ -4585,6 +4591,7 @@ static void doInitCacheIds() {
 	 * not done (we're in the process of initializing the caching mechanisms).
 	 * We must ensure the classes that own the fields of interest are prepared.
 	 */
+	/*[IF JAVA_SPEC_VERSION == 8]*/
 	J9VMInternals.prepare(Constructor.class);
 	J9VMInternals.prepare(Method.class);
 	try {
@@ -4596,12 +4603,28 @@ static void doInitCacheIds() {
 	constructorParameterTypesField.setAccessible(true);
 	methodParameterTypesField.setAccessible(true);
 	if (reflectCacheEnabled) {
+		J9VMInternals.prepare(Executable.class);
+		J9VMInternals.prepare(Field.class);
 		copyConstructor = getAccessibleMethod(Constructor.class, "copy");
 		copyMethod = getAccessibleMethod(Method.class, "copy");
 		copyField = getAccessibleMethod(Field.class, "copy");
+		getRootMethod = getAccessibleMethod(Executable.class, "getRoot");
+		try {
+			rootField = Field.class.getDeclaredFieldImpl("root");
+		} catch (NoSuchFieldException e) {
+			throw newInternalError(e);
+		}
+		rootField.setAccessible(true);
 	}
+	/*[ELSE] JAVA_SPEC_VERSION == 8 */
+	J9VMInternals.prepare(AccessibleObject.class);
+	if (reflectCacheEnabled) {
+		getRootMethod = getAccessibleMethod(AccessibleObject.class, "getRoot");
+	}
+	/*[ENDIF] JAVA_SPEC_VERSION == 8 */
 }
-/*[ENDIF] JAVA_SPEC_VERSION == 8 */
+/*[ENDIF] JAVA_SPEC_VERSION <= 11 */
+
 private static Method getAccessibleMethod(Class<?> cls, String name) {
 	try {
 		Method method = cls.getDeclaredMethod(name, EmptyParameters);
@@ -4974,16 +4997,24 @@ private static final class ReflectCache extends ConcurrentHashMap<CacheKey, Refl
 		return ref != null ? ref.get() : null;
 	}
 
-	void insert(CacheKey key, Object value) {
+	void insert(CacheKey key, AccessibleObject value) {
+		insertRoot(key, retrieveRoot(value));
+	}
+
+	/*
+	 * This method must only be called with known roots or aggregates of roots.
+	 */
+	void insertRoot(CacheKey key, Object value) {
 		put(key, new ReflectRef(this, key, value));
 	}
 
-	<T> T insertIfAbsent(CacheKey key, T value) {
-		ReflectRef newRef = new ReflectRef(this, key, value);
+	<T extends AccessibleObject> T insertIfAbsent(CacheKey key, T value) {
+		T rootValue = retrieveRoot(value);
+		ReflectRef newRef = new ReflectRef(this, key, rootValue);
 		for (;;) {
 			ReflectRef oldRef = putIfAbsent(key, newRef);
 			if (oldRef == null) {
-				return value;
+				return rootValue;
 			}
 			T oldValue = (T) oldRef.get();
 			if (oldValue != null) {
@@ -4992,8 +5023,20 @@ private static final class ReflectCache extends ConcurrentHashMap<CacheKey, Refl
 			// The entry addressed by key has been cleared, but not yet removed from this map.
 			// One thread will successfully replace the entry; the value stored will be shared.
 			if (replace(key, oldRef, newRef)) {
-				return value;
+				return rootValue;
 			}
+		}
+	}
+
+	private static <T extends AccessibleObject> T retrieveRoot(T value) {
+		T root = getRoot(value);
+		if (root != null) {
+			if (reflectCacheDebug) {
+				System.err.println("extracted root from non-root accessible object before caching: " + value);
+			}
+			return root;
+		} else {
+			return value;
 		}
 	}
 
@@ -5120,7 +5163,7 @@ private Method cacheMethod(Method method) {
 		try {
 			// cache the Method with the largest depth with a null returnType
 			CacheKey lookupKey = CacheKey.newMethodKey(method.getName(), parameterTypes, null);
-			cache.insert(lookupKey, method);
+			cache.insertRoot(lookupKey, method);
 		} finally {
 			cache.release();
 		}
@@ -5188,7 +5231,7 @@ private Field cacheField(Field field) {
 		if (declaringClass == this) {
 			// cache the Field returned from getField() with a null returnType
 			CacheKey lookupKey = CacheKey.newFieldKey(field.getName(), null);
-			cache.insert(lookupKey, field);
+			cache.insertRoot(lookupKey, field);
 		}
 	} finally {
 		cache.release();
@@ -5349,7 +5392,7 @@ private Method[] cacheMethods(Method[] methods, CacheKey cacheKey) {
 		if (cache == null) {
 			cache = acquireReflectCache();
 		}
-		cache.insert(cacheKey, methods);
+		cache.insertRoot(cacheKey, methods);
 	} finally {
 		if (cache != null) {
 			cache.release();
@@ -5433,7 +5476,7 @@ private Field[] cacheFields(Field[] fields, CacheKey cacheKey) {
 		if (cache == null) {
 			cache = acquireReflectCache();
 		}
-		cache.insert(cacheKey, fields);
+		cache.insertRoot(cacheKey, fields);
 	} finally {
 		if (cache != null) {
 			cache.release();
@@ -5499,7 +5542,7 @@ private Constructor<T>[] cacheConstructors(Constructor<T>[] constructors, CacheK
 			CacheKey key = CacheKey.newConstructorKey(getParameterTypes(constructors[i]));
 			constructors[i] = cache.insertIfAbsent(key, constructors[i]);
 		}
-		cache.insert(cacheKey, constructors);
+		cache.insertRoot(cacheKey, constructors);
 	} finally {
 		cache.release();
 	}
@@ -5546,6 +5589,30 @@ Object setMethodHandleCache(Object cache) {
 		}
 	}
 	return result;
+}
+
+static <T extends AccessibleObject> T getRoot(T object) {
+	/*[IF JAVA_SPEC_VERSION >= 17]*/
+	return SharedSecrets.getJavaLangReflectAccess().getRoot(object);
+	/*[ELSEIF JAVA_SPEC_VERSION >= 11] */
+	try {
+		return (T) getRootMethod.invoke(object, NoArgs);
+	} catch (IllegalAccessException | InvocationTargetException e) {
+		throw newInternalError(e);
+	}
+	/*[ELSE] JAVA_SPEC_VERSION >= 11 */
+	try {
+		if (object instanceof Executable) {
+			return (T) getRootMethod.invoke(object, NoArgs);
+		} else if (object instanceof Field) {
+			return (T) rootField.get(object);
+		} else {
+			throw newInternalError(new IllegalArgumentException("Unexpected AccessibleObject subclass: " + object.getClass().getSimpleName()));
+		}
+	} catch (IllegalAccessException | InvocationTargetException e) {
+		throw newInternalError(e);
+	}
+	/*[ENDIF] JAVA_SPEC_VERSION >= 17 */
 }
 
 ConstantPool getConstantPool() {

